### DISCUSSION TITLE
[Question] Necessary to check IsReplaying before SetCustomStatus?

### DISCUSSION THREAD

**Discussion by colbylwilliams** [Azure/azure-functions-durable-extension#1764] [Q&A] [answered]
### Question

Is it necessary to check if an orchestration function is replaying before setting a custom status?  For example:

``` csharp
if (!functionContext.IsReplaying)
    functionContext.SetCustomStatus("Creating Project");

var _ = await functionContext
      .CallActivityAsync<Guid>(nameof(CreateProjectActivity), project);

if (!functionContext.IsReplaying)
    functionContext.SetCustomStatus("Creating Project Users");

var _ = await functionContext
      .CallActivityAsync<Guid>(nameof(CreateProjectUsersActivity), project);

functionContext.SetCustomStatus("Project Created");
```
Or is this handled automatically?  I'd assume replays are fast, but I'm not sure how long it takes for the `CustomStatus` to be persisted and available via the HTTP APIs.  So if I don't check `IsReplaying` is there an (edge) chance that the HTTP API could return `"Creating Project"` even though `CreateProjectUsersActivity` is really already finished?  (i.e. the HTTP API was queried mid-replay)

**Comment by cgillum** [comment, marked as answer]
It's not necessary to check for `IsReplaying`.

The `SetCustomStatus` API doesn't actually do any I/O at the time you call it. Rather, it just updates some in-memory state. The I/O happens at the point where we checkpoint the updated instance history (usually the next `await` statement). Only the data saved in the last `SetCustomStatus` call will get persisted. This also means that you'll never see a "Creating Project" status when querying the instance status, even if the query happens mid-replay.

Does that help clarify?

**Comment by colbylwilliams** [comment]
Yep makes sense.  Thanks @cgillum
